### PR TITLE
fix: Fixes issue with polyfills overriding native shadowDOM functions

### PR DESCRIPTION
**Pull request by lukebatchelor**

### CODE REVIEW

File: src/api/symbols.js [hunk 1]
```diff
@@ -5,3 +5,5 @@ export const props = '____props';
 export const renderer = '____renderer';
 export const rendererDebounced = '____rendererDebounced';
 export const shadowRoot = '____shadowRoot';
+export const v0ShadowDOMProperty = 'createShadowRoot'; // property used to check for shadowDOMv0 support
```
Comment (treshugart) [inline src/api/symbols.js]: Eventually these symbols will actually use `Symbol()`, which you wouldn't want to do with these. I'd suggest just co-locating these in the support file.

File: src/util/support.js [hunk 2]
```diff
@@ -1,5 +1,36 @@
+import isNativeRegex from 'regex-native-function'; // checks if a function is native using regex
```
Comment (treshugart) [inline src/util/support.js]: You may not need this. You can check for native v0 by doing `'HTMLContentElement' in window` and v1 by doing `'HTMLSlotElement' in window`.

Comment (nadiam84) [inline src/util/support.js]: webcomponents polyfilling HTMLContentElement

Comment (lukebatchelor) [inline src/util/support.js]: Yea, we were trying to make sure that the check wont return true if a polyfill had set the property already. Unless there's another reason not to use the regex way?

Comment (treshugart) [inline src/util/support.js]: Not really, I was just exploring built-in ways to do this. Just trying to keep deps to a minimum, if possible.

File: src/util/support.js [hunk 3]
```diff
@@ -1,5 +1,36 @@
...
+import { v0ShadowDOMProperty, v1ShadowDOMProperty } from '../api/symbols';
+
 const div = document.createElement('div');
+const isNative = (elem, prop) => isNativeRegex.test(elem[prop]);
+const isPolyfill = (elem, prop) => !!elem[prop] && !isNative(elem, prop);
```
Comment (treshugart) [inline src/util/support.js]: Should the regex return false if `elem[prop]` is not defined? If so, this would mean you don't need that first check, right?

Comment (nadiam84) [inline src/util/support.js]: Yes, but it also will return `false` if this property was polyfilled.

Comment (treshugart) [inline src/util/support.js]: Saw that just after I commented :)

File: README.md [hunk 4]
```diff
@@ -60,6 +60,7 @@ Skate requires Custom Element support. In [browsers that don't support it](http:
 
 Skate prefers v0 support if it detects both as that yields the best performance in browsers that support it natively. If you supply only a v1 polyfill, it will use v0 where supported by native and v1 in other browsers.
 
+Skate will always prefer a native implementation to a polyfill as they provide the best performance. Fo example, if you had native v0 support and v1 support through a polyfill, Skate will use v0. Likewise, if native support was detected for both, Skate would use v1.
```
Comment (nadiam84) [inline README.md]: 'for'

File: test/unit/util/support.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import { shouldUseShadowDomV0, shouldUseShadowDomV1, v0ShadowDOMProperty, v1ShadowDOMProperty } from '../../../src/util/support';
+
+describe('util/support', () => {
+  describe('shadowDOM', () => {
+    /* We take a copy of a known native property so that we can mock how our code will behave when 
+       it finds a native implmentation of a function. The polyfilled property just needs to match
```
Comment (nadiam84) [inline test/unit/util/support.js]: 'implementation'

File: README.md (hunk 4 above)
Comment (lukebatchelor) [inline README.md]: Nuh, I'm speaking Gangsta'

File: test/unit/util/support.js (hunk 5 above)
Comment (lukebatchelor) [inline test/unit/util/support.js]: Nuh, I'm speaking... ..

I've got nothing.

Comment (nadiam84) [inline test/unit/util/support.js]: :D :D

File: src/util/support.js [hunk 6]
```diff
@@ -1,5 +1,38 @@
+import isNativeRegex from 'regex-native-function'; // checks if a function is native using regex
+
 const div = document.createElement('div');
+const isNative = (elem, prop) => isNativeRegex.test(elem[prop]);
+const isPolyfill = (elem, prop) => !!elem[prop] && !isNative(elem, prop);
+
+export const v0ShadowDOMProperty = 'createShadowRoot'; // property used to check for shadowDOMv0 support
```
Comment (treshugart) [inline src/util/support.js]: nit: comments go above associated lines, empty newline should separate comment and code above it, i.e.:

```js
// comment 1
const code1;

// comment 2
const code2;
```

The aim is readability and comment association.

File: test/unit/util/support.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
+import { shouldUseShadowDomV0, shouldUseShadowDomV1, v0ShadowDOMProperty, v1ShadowDOMProperty } from '../../../src/util/support';
+
+describe('util/support', () => {
+  describe('shadowDOM', () => {
+    /* We take a copy of a known native property so that we can mock how our code will behave when
```
Comment (treshugart) [inline test/unit/util/support.js]: nit: multiline comments should still use `//`

Comment (lukebatchelor) [inline test/unit/util/support.js]: So just use /**/ for headers and jsdoc annotations and stuff?

**Comment by treshugart** [comment]
lgtm

**Comment by lukebatchelor** [comment]
Can I LGTM?

**Pull request merged**